File: src/main/java/net/loxal/soa/restkit/client/RESTClient.java
/*
 * Copyright 2014 Alexander Orlov <alexander.orlov@loxal.net>. All rights reserved.
 */

package net.loxal.soa.restkit.client;

import javax.ws.rs.client.Client;
import javax.ws.rs.client.ClientBuilder;
import javax.ws.rs.client.Entity;
import javax.ws.rs.client.Invocation;
import javax.ws.rs.client.WebTarget;
import javax.ws.rs.core.Response;
import java.util.logging.Logger;

public abstract class RESTClient<T> {
    static final Logger LOG = Logger.getGlobal();
    static final Client CLIENT = ClientBuilder.newClient();

    Invocation.Builder targetTenant(WebTarget webTarget) {
        return webTarget.request();
    }

    public abstract Response post(Entity<T> entity);

    public abstract Response delete(Class<T> entityType, String id);

    public abstract Response get(Class<T> entityType, String id);

    public abstract Response put(Entity<T> json, String id);
}

File: src/main/kotlin/net/loxal/soa/restkit/client/RestClient.kt
/*
 * Copyright 2015 Alexander Orlov <alexander.orlov@loxal.net>. All rights reserved.
 */

package net.loxal.soa.restkit.client

import org.slf4j.LoggerFactory
import javax.ws.rs.client.ClientBuilder
import javax.ws.rs.client.Entity
import javax.ws.rs.client.Invocation
import javax.ws.rs.client.WebTarget
import javax.ws.rs.core.HttpHeaders
import javax.ws.rs.core.Response

abstract class RestClient<out T> protected constructor() {

    fun targetTenant(webTarget: WebTarget): Invocation.Builder = webTarget.request()

    fun applyToken(request: Invocation.Builder): Invocation.Builder = request.header(HttpHeaders.AUTHORIZATION, "Bearer ${RepositoryClient.authorization.access_token}")

    fun authorizeRequest(target: WebTarget): Invocation.Builder = applyToken(targetTenant(target))

    abstract fun post(entity: Entity<in T>, id: String): Response

    abstract fun delete(entityType: Class<in T>, id: String): Response

    abstract fun get(entityType: Class<in T>, id: String): Response

    abstract fun put(json: Entity<in T>, id: String): Response

    companion object {
        val LOG = LoggerFactory.getLogger(RestClient::class.java)
        val CLIENT = ClientBuilder.newClient()
    }
}